module/VuFind/src/VuFind/Controller/AbstractSearch.php

```diff
@@ -374,6 +375,7 @@ protected function getSearchResultsView($setupCallback = null)
         $lastView = $this->getSearchMemory()
             ->retrieveLastSetting($this->searchClassId, 'view');
         try {
+            /** @var Results $results */
```

Comment:
I'm guessing that your IDE may be injecting type annotations automatically. Do you mind turning that off? While I definitely want to improve the type specificity and accuracy in the code base, I think we're going to want to do that in a slightly different way than what I'm seeing in this PR, and these annotations may cause confusion. (If I'm mistaken about how I'm interpreting what I see here and you're putting these in intentionally rather than having them added automatically by accident, I'm certainly open to discussing further).

Reply:
I put them intentionally, in order to use all the features from the IDE (code completion, refactoring, navigation, ...) (I use PHPStorm).
In this case the IDE can not resolve what's being done in the function and what's returned and hence can not provide all the features
I don't mind removing them, and I'm also open to suggestion/discussion how to implement them but I don't want to revolutionize how the project is done

Reply:
I think there's a fairly simple change we can make that will greatly improve type interpretation. I would add this method to \VuFind\Controller\AbstractBase.php:

```
    /**
     * Retrieve a service
     *
     * @param class-string<T> $name Name of service to retrieve
     *
     * @template T
     *
     * @return T
     */
    public function getService(string $name)
    {
        return $this->serviceLocator->get($name);
    }
```

And then replace all of the calls to `$this->serviceLocator->get()` in our controllers with `$this->getService()`.

That's not going to solve every problem, but it should solve the one you're trying to address with annotations here. Obviously, this change is out of scope for the present PR, but if you'd like to open a separate one to try it out, I'd be open to merging it.

(I should also say that in the long term, the goal is to move more toward dependency injection and away from service locator magic... but in the short term, having better types on the existing code would be beneficial).

Reply:
I added the piece of code you mentioned and I removed the annotation above, however I didn't remove the annotations in the view, do you have a preference concerning those?

Reply:
Thanks! I don't think it's necessary to remove the annotations if they are useful. In the long term I'd like a way to enforce these consistently, since it bothers me a little to have them in some places but not others... but we have to start somewhere! I'd suggest perhaps boosting them all the way to the tops of templates where they are used, though -- I think it may improve readability to declare variables in one spot prior to the code, similar to use declarations in other files.